src/plasTeX/Renderers/__init__.py

```diff
@@ -295,7 +295,7 @@ def url_fragment(self):
         """
         The ``id`` of this node, suitable for use in a url fragment.
         """
-        return url_quote(self.id)
+        return urllib_parse.quote(self.id.encode('utf-8'))
```

Comment:
Two things:

1. This should be something like `self.id.encode('encoding') if not isinstance(self.id, str) else self.id`. Otherwise, for the case that it's already a str, Python 2 could raise a UnicodeError because it will implicitly *decode* it first to unicode using 'ascii'. And in all cases, if it's already a str it will be doing too much work.
2. I'm not sure this is actually what you want, it won't produce very usable URL fragments:
```pycon
Python 2.7.13 (84a2f3e6a7f8, Oct 03 2017, 18:46:00)
[PyPy 5.9.0 with GCC 4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.40)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import urllib
>>> urllib.quote(u'\u1f600')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jmadden/bin/pypy/lib-python/2.7/urllib.py", line 1300, in quote
    return ''.join(map(quoter, s))
KeyError: u'\U0001f600'
>>> urllib.quote(u'\u1f600'.encode("utf-8"))
'%F0%9F%98%80'
```
Somewhere there's a library we use for transliteration of non-latin languages to their closest latin equivalent, ideally we'd run it through that. Otherwise, we should probably do something like:
```pycon
>>> urllib.quote(u'\u1f600'.encode("latin-1", 'replace').replace('?', 'x'))
'x'
```

Reply:
We are getting KeyError: u'\u2013' when creating a section node from an rST node. The string in question is u'The Failure of Reconstruction, 1870\u20131877'.  The string is already unicode in py2.

Reply:
We can encode -latin replace in code that calls plasTeX

Reply:
Right. On Python 2, urllib.quote only deals with str---any chars that need quoting have to already be a str.
```pycon
Python 2.7.13 (84a2f3e6a7f8, Oct 03 2017, 18:46:00)
[PyPy 5.9.0 with GCC 4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.40)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import urllib
>>> urllib.quote(u'unicode')
u'unicode'
>>> urllib.quote(b'bytes')
'bytes'
>>> urllib.quote(u'a space')
'a%20space' # Notice this became bytes
>>> urllib.quote(u'\u1f600')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jmadden/bin/pypy/lib-python/2.7/urllib.py", line 1300, in quote
    return ''.join(map(quoter, s))
KeyError: u'\U0001f600'
```
Only in Python 3 can quote handle both str and bytes.

Reply:
> We can encode -latin replace in code that calls plasTeX

I do think plasTeX can do better here. It needs another good look to carefully define what's supposed to be text and what's supposed to be bytes and deal appropriately. Our test coverage isn't amazing so there are probably lots of things like that lurking where implicit conversions happen (on Py2) but would error on Py3 or with unexpected input on Py2.